features/steps/steps.py

```diff
@@ -75,9 +75,20 @@ def __str__(self):
 class instance_structure_error:
     related: ifcopenshell.entity_instance
     relating: ifcopenshell.entity_instance
+    relationship_type: str
+    optional_values: field(default_factory=dict)
 
     def __str__(self):
-        return f"The instance {fmt(self.related)} is assigned to {fmt(self.relating)}"
+        self.pos_neg = 'is not' if do_try(lambda: self.optional_values['condition'], '') == 'must' else 'is'
```

Comment:
This is a bit "muggenziften", but try not to modify `self` in `__str__`-like functions. It's not really a concern here, but you wouldn't expect. Below is a silly example, but hopefully it demonstrates how surprising it is when functions like `__str__()` affect the state (they are called side effects). In other languages you would expect these functions to be marked `const` so they can't affect state. Python doesn't have that. In short: just define local variables `pos_neg` `directness` don't modify self.

```
>>> class A:
...   value = 1
...   def __str__(self):
...     v = str(self.value)
...     self.value *= 2
...     return v
...
>>>
>>>
>>> a = A()
>>>
>>> a.value
1
>>> print(a)
1
>>> a.value
2
>>> print(a)
2
>>> print(a)
4
```

Reply:
That is good to know, I would not expect that indeed. It is modified for now and you probably will save a future-me some frustrations when this occurs on another occasion.